### Coda

The tiny core of the Coda system is contained in the base module.  To run this demo, your PYTHONPATH should include the .../py directory of coda. 

In [1]:
from base import * 

By definition, data is a finite sequence of data. An empty data is created in python by `data()`.  This can be displayed either as `()` or as a symbol representing the empty set in python by using `repr(data())` or `str(data())` respectively. 

In [2]:
#
#  create an empty pure data and display it
#
z = data()
print(repr(z),str(z))

() ∅


The data constructor takes tuples of data as arguments. 

In [3]:
zz = data(z,z,z,z)
print(repr(zz),zz)

(()()()()) (∅∅∅∅)


In addition to the data constructor, there are two foundational operations from data to data. If x and y are data, the operations are

1. Concatenation, indicated by x y in the language or by x+y in python.
2. Colon, indicated by x:y in the language or by x|y in python. 

As you can see, all data are merely recursive sequences of empty sequences.

If z is the empty data, the 0 and 1 bits are defined to be z|z and z|(z|z) respectively.  Characters are data consisting of sequences of these, words are sequences of characters, etc. 

In [4]:
L = [z,z+z,data(z,z),z|z,z|z|z,z|(z|z),data(z|z,z|z|z,(z|(z|z)))]
for l in L: print(l,repr(l),hash(l))

∅ () 5740354900026072187
∅ () 5740354900026072187
𝟭 (()()) 9028247024705308198
𝟬 (()) -5486347211504344842
(𝟬) ((())) 4510597632111149919
𝟭 (()()) 9028247024705308198
(𝟬(𝟬)𝟭) ((())((()))(()())) 7764245749913446555


In [5]:
print(co('Hello').type())

𝟬


In [6]:
#
#  The co function creates pure data from unicode strings in python
#
co('Hello World')

(((())(()())(())(())(()())(())(())(()))((())(()())(()())(())(())(()())(())(()()))((())(()())(()())(())(()())(()())(())(()))((())(()())(()())(())(()())(()())(())(()))((())(()())(()())(())(()())(()())(()())(()()))((())(())(()())(())(())(())(())(()))((())(()())(())(()())(())(()())(()())(()()))((())(()())(()())(())(()())(()())(()())(()()))((())(()())(()())(()())(())(())(()())(()))((())(()())(()())(())(()())(()())(())(()))((())(()())(()())(())(())(()())(())(())))

In [7]:
print(co('Hello World'))

(Hello World)


In [8]:
#
#   Pure data is merely a sequence of data.  You can access this sequence in python as if it is a tuple.
#
H = co('Hello World')
print(H[6],'<-',repr(H[6]))
for h in H[6]: print(h,'<-',repr(h))

W <- ((())(()())(())(()())(())(()())(()())(()()))
𝟬 <- (())
𝟭 <- (()())
𝟬 <- (())
𝟭 <- (()())
𝟬 <- (())
𝟭 <- (()())
𝟭 <- (()())
𝟭 <- (()())


In [9]:
#
#   Bits are pure data also, so everything is made of empty sequences. 
#
one = H[6][1]
print(one,'<-',repr(one))
for d in one: print(d,'<-',repr(d))

𝟭 <- (()())
∅ <- ()
∅ <- ()


For more examples, we can use the Universe module to create all data with specified width and depth less than specified values. The 3,2 universe, for example is all data with width <= 3 and depth <= 2.  

In [10]:
#
#   The 1,1 universe contains only empty and the zero bit. 
#
import Universe
for d in Universe.universe(1,1): print(repr(d),' -> ',d)

()  ->  ∅
(())  ->  𝟬


In [11]:
#
#   The 2,1 universe also contains binary 1. 
#
for d in Universe.universe(2,1): print(repr(d),' -> ',d)

()  ->  ∅
(())  ->  𝟬
(()())  ->  𝟭


In [12]:
U = Universe.universe(3,2)
for d in U: print(d)

∅
𝟬
𝟭
(∅∅∅)
(𝟬)
(𝟭)
((∅∅∅))
(𝟬𝟬)
(𝟬𝟭)
(𝟬(∅∅∅))
(𝟭𝟬)
(𝟭𝟭)
(𝟭(∅∅∅))
((∅∅∅)𝟬)
((∅∅∅)𝟭)
((∅∅∅)(∅∅∅))
(𝟬𝟬𝟬)
(𝟬𝟬𝟭)
(𝟬𝟬(∅∅∅))
(𝟬𝟭𝟬)
(𝟬𝟭𝟭)
(𝟬𝟭(∅∅∅))
(𝟬(∅∅∅)𝟬)
(𝟬(∅∅∅)𝟭)
(𝟬(∅∅∅)(∅∅∅))
(𝟭𝟬𝟬)
(𝟭𝟬𝟭)
(𝟭𝟬(∅∅∅))
(𝟭𝟭𝟬)
(𝟭𝟭𝟭)
(𝟭𝟭(∅∅∅))
(𝟭(∅∅∅)𝟬)
(𝟭(∅∅∅)𝟭)
(𝟭(∅∅∅)(∅∅∅))
((∅∅∅)𝟬𝟬)
((∅∅∅)𝟬𝟭)
((∅∅∅)𝟬(∅∅∅))
((∅∅∅)𝟭𝟬)
((∅∅∅)𝟭𝟭)
((∅∅∅)𝟭(∅∅∅))
((∅∅∅)(∅∅∅)𝟬)
((∅∅∅)(∅∅∅)𝟭)
((∅∅∅)(∅∅∅)(∅∅∅))


In [13]:
#
#   The sizes of universes grows very rapidly.
#
print(len(Universe.universe(3,3)))

60922


## Universe Sizes for given widths and depths 

| Depth     | 1  |  2  |  3    |  4  |      5  |
|----------:|---:|----:|------:|----:|--------:|
| Width 1:   | 2  |  3  |  4    |  5  |      6  |
| Width 2:   | 3  |  9  | 51    | 1857| 3265299 | 
| Width 3:   | 4  | 43  | 60922 | 
| Width 4:   | 5  | 345 | 
| Width 5:   | 6  | 3911| 
| Width 6:   | 7  | 55993 |
| Width 7:   | 8  | 960807 | 